scalding-core/src/main/scala/com/twitter/scalding/typed/cascading_backend/AsyncFlowDefRunner.scala

```diff
@@ -242,17 +255,20 @@ class AsyncFlowDefRunner(mode: CascadingMode) extends Writer {
    * calls runFlowDef, then clears the FlowStateMap
    */
   def validateAndRun(conf: Config)(fn: Config => FlowDef)(
-    implicit cec: ConcurrentExecutionContext): Future[(Long, ExecutionCounters)] =
-    for {
+    implicit cec: ConcurrentExecutionContext): (Future[(Long, ExecutionCounters)], CancellationHandler) = {
+    val fut = for {
       flowDef <- Future(fn(conf))
       _ = FlowStateMap.validateSources(flowDef, mode)
-      (id, jobStats) <- runFlowDef(conf, flowDef)
+      (resultFut, cancelHandler) = runFlowDef(conf, flowDef) // how to pass cancelHandler through?
+      (id, jobStats) <- resultFut
       _ = FlowStateMap.clear(flowDef)
-    } yield (id, ExecutionCounters.fromJobStats(jobStats))
+    } yield ((id, ExecutionCounters.fromJobStats(jobStats)), cancelHandler)
+    (fut.map(_._1), CancellationHandler.empty.compose(fut.map(_._2)))
```

Comment:
rather than make a `compose` that takes `Future[CancellationHandler]` maybe just have a method like `def handleFromFuture(f: Future[CancelationHandle]): CancelationHandle`, which I think is what you want here.

I think the way you have done it is fine, although I agree it is awkward to lift out of the Future.

Reply:
👍